ethexe/validator/src/lib.rs

```diff
@@ -16,99 +16,437 @@
-        !self.current_blocks.is_empty()
-    }
-
     pub fn pub_key(&self) -> PublicKey {
         self.pub_key
     }
 
-    pub fn codes_aggregation(&mut self) -> Result<AggregatedCommitments<CodeCommitment>> {
+    pub fn address(&self) -> Address {
+        self.pub_key.to_address()
+    }
+
+    pub fn aggregate<C: ToDigest>(&self, commitments: Vec<C>) -> Result<AggregatedCommitments<C>> {
         AggregatedCommitments::aggregate_commitments(
-            self.current_codes.clone(),
+            commitments,
             &self.signer,
             self.pub_key,
             self.router_address,
         )
     }
 
-    pub fn blocks_aggregation(&mut self) -> Result<AggregatedCommitments<BlockCommitment>> {
-        AggregatedCommitments::aggregate_commitments(
-            self.current_blocks.clone(),
+    pub fn validate_code_commitments(
+        &mut self,
+        db: &impl CodesStorage,
+        requests: impl IntoIterator<Item = CodeCommitment>,
+    ) -> Result<(Digest, Signature)> {
+        let mut commitment_digests = Vec::new();
+        for request in requests {
+            commitment_digests.push(request.to_digest());
+            Self::validate_code_commitment(db, request)?;
+        }
+
+        let commitments_digest = commitment_digests.to_digest();
+        agro::sign_commitments_digest(
+            commitments_digest,
             &self.signer,
             self.pub_key,
             self.router_address,
         )
+        .map(|signature| (commitments_digest, signature))
     }
 
-    pub fn push_commitments(&mut self, commitments: Vec<Commitment>) -> Result<()> {
-        for commitment in commitments {
-            match commitment {
-                Commitment::Code(code_commitment) => self.current_codes.push(code_commitment),
-                Commitment::Block(block_commitment) => self.current_blocks.push(block_commitment),
-            }
+    pub fn validate_block_commitments(
+        &mut self,
+        db: &impl BlockMetaStorage,
+        requests: impl IntoIterator<Item = BlockCommitmentValidationRequest>,
+    ) -> Result<(Digest, Signature)> {
+        let mut commitment_digests = Vec::new();
+        for request in requests.into_iter() {
+            commitment_digests.push(request.to_digest());
+            Self::validate_block_commitment(db, request)?;
         }
 
+        let commitments_digest = commitment_digests.to_digest();
+        agro::sign_commitments_digest(
+            commitments_digest,
+            &self.signer,
+            self.pub_key,
+            self.router_address,
+        )
+        .map(|signature| (commitments_digest, signature))
+    }
+
+    fn validate_code_commitment(db: &impl CodesStorage, request: CodeCommitment) -> Result<()> {
+        let CodeCommitment { id: code_id, valid } = request;
+        if db
+            .code_valid(code_id)
+            .ok_or(anyhow!("Code {code_id} is not validated by this node"))?
+            .ne(&valid)
```

Comment:
why not just use the operator directly? this also applies to .not()

Reply:
Because `cargo fmt` makes very stupid code lines alignment, for example
```rust
        if db.block_prev_commitment(block_hash).ok_or(anyhow!(
            "Cannot get from db previous commitment block for block {block_hash}"
        ))? != allowed_prev_commitment_hash
        {
        }
```

Reply:
Done